### ISSUE TITLE
Version 2.37 dead code elimination broken on Rambda ES curry

### ISSUE THREAD

**Issue by StreetStrider** [rollup/rollup#3942]
- Rollup Version: 2.37
- Operating System (or Browser): GNU/Linux
- Node Version (if applicable): v12.19.1
- Link to reproduction (IMPORTANT, read below): [rollup-dce-curry.zip](https://github.com/rollup/rollup/files/5898497/rollup-dce-curry.zip)

Hello! I've may tracked down dead code elimination eliminating the body of usual curry function. This leaves all curried functions broken. I've prepped test case where 2.36 works OK. Then, updating rollup to 2.37 draws the issue.

### Expected Behavior
Update Rollup in test case to 2.37. Curried functions works as intended.

### Actual Behavior
`<function name> is not a function`

**Comment by kzc** [comment]
The issue is related to module `sideEffects`.

Given:

```
$ rollup -v
rollup v2.38.1
```
```
$ cat node_modules/rambda/package.json 
{
  "sideEffects": false
}
```
```
$ cat node_modules/rambda/src/curry.js 
export function curry(fn, args = []){
  return (..._args) =>
    (rest => rest.length >= fn.length ? fn(...rest) : curry(fn, rest))([
      ...args,
      ..._args,
    ])
}
```
```
$ cat input.js 
import {curry} from './node_modules/rambda/src/curry'
function foo() { console.log('Foo') }
let c = curry(foo);
c();
```
Correct output with no plugins:
```
$ rollup input.js --silent
function curry(fn, args = []){
  return (..._args) =>
    (rest => rest.length >= fn.length ? fn(...rest) : curry(fn, rest))([
      ...args,
      ..._args,
    ])
}

function foo() { console.log('Foo'); }
let c = curry(foo);
c();
```
Incorrect output with the node-resolve plugin:
```
$ rollup input.js --silent -p node-resolve
function curry(fn, args = []){
}

function foo() { console.log('Foo'); }
let c = curry(foo);
c();
```

**Comment by lukastaegert** [comment]
Thanks for the triage, fix at #3944
